### PR TITLE
docs: kernel policy support

### PR DESCRIPTION
**Pull request by dianpopa**
Added the policy behind validating host/guest
kernel versions with Firecracker.

Signed-off-by: Diana Popa <dpopa@amazon.com>

# Reason for This PR

`[Author TODO: add issue #.]`
`[Open the PR after the related issue has a clear conclusion.]`
`[If there is no issue which states the need for this PR, create one first.]`
Fixes #

## Description of Changes

`[Author TODO: add description of changes.]`

- [ ] This functionality can be added in [`rust-vmm`][1].

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The issue which led to this PR has a clear conclusion.
- [ ] This PR follows the solution outlined in the related issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes follow the [Runbook for Firecracker API changes][2].
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

**Comment by alindima** [comment]
I think we should we have a link to this document somewhere in the README, WDYT?
I'd also add a Changelog update to specifically state that we added a support policy and a link to it.

**Comment by raduiliescu** [comment]
I wouldn't add this in the changelog since is not a per release thing. In the changelog I would add something like add support and verification for kernel 5.10.
Beside the readme, I would review the docs and add links to this in the places where kernel and kernel operations are mentioned, if it makes sense.

File: docs/kernel-policy.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Firecracker's Kernel Support Policy
+
+Being a virtual machine monitor, Firecracker represents a component in a
+larger stack, one in which it is tightly coupled with the guest and host
+kernels on which it is run.
+
+As of right now, Firecracker is being validated
```
Comment (raduiliescu) [inline docs/kernel-policy.md:7]: I would say "Firecracker code from main branch is being validated"

File: docs/kernel-policy.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+continuously on *4.14* and *5.10* host and guest kernels. While other
+versions and other kernel configs might work, they are not periodically
+validated in our test suite, and using them might result in unexpected behaviour.
+Once enabled, a kernel version is supported for a **minimum of 2 years**.
+
+We are validating the currently supported Firecracker releases as per
+[Firecracker’s release policy](../docs/RELEASE_POLICY.md). Adding support
```
Comment (raduiliescu) [inline docs/kernel-policy.md:14]: I would add before adding support... Starting with release v1.0 each major and minor release will specify the supported kernel versions. Subsequent patch releases will continue to support the same kernel versions as the minor release.

File: docs/kernel-policy.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# Firecracker's Kernel Support Policy
+
+Being a virtual machine monitor, Firecracker represents a component in a
+larger stack, one in which it is tightly coupled with the guest and host
+kernels on which it is run.
```
Comment (raduiliescu) [inline docs/kernel-policy.md:5]: I  would add that With this kernel support policy we want to offer customers predictability into future kernel related changes.

File: docs/kernel-policy.md (hunk 2 above)
Comment (georgepisaltu) [reply docs/kernel-policy.md:14]: Shouldn't this be mentioned in the release policy rather than here?

File: docs/kernel-policy.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+Below is a per-functionality breakdown of guest kernel modules
+relevant to Firecracker for all platforms:
+
+* serial console - `CONFIG_SERIAL_8250_CONSOLE`, `CONFIG_PRINTK`
+* initrd support - `CONFIG_BLK_DEV_INITRD`
+* virtio devices - `CONFIG_VIRTIO_MMIO`
+  * balloon - `CONFIG_MEMORY_BALLOON`, `CONFIG_VIRTIO_BALLOON`
+  * block - `CONFIG_VIRTIO_BLK`
+    * partuuuid support - `CONFIG_MSDOS_PARTITION`
```
Comment (georgepisaltu) [inline docs/kernel-policy.md:51]:
```suggestion
    * partuuid support - `CONFIG_MSDOS_PARTITION`
```

File: docs/kernel-policy.md (hunk 2 above)
Comment (raduiliescu) [reply docs/kernel-policy.md:14]: Is a good idea. I propose to mention it there.

**Comment by dianpopa** [comment]
I added references to the kernel support policy throughout our doc.
Also mentioned it in the CHANGELOG.
@raduiliescu @alindima @georgepisaltu let me know what you think!

File: FAQ.md [hunk 5]
```diff
@@ -70,10 +70,10 @@ and guest.
 
 ### What operating systems are supported by Firecracker?
 
-Firecracker supports Linux host and guest operating systems with kernel versions
-4.14 and above, as well as
+Firecracker supports Linux host and guest operating systems as well as
 [OSv](http://blog.osv.io/blog/2019/04/19/making-OSv-run-on-firecraker/) guests.
-The long-term support plan is still under discussion.
+Currently supported host/guest kernel versions can be found in the
```
Comment (raduiliescu) [inline FAQ.md:75]: I suggest the following change: Firecracker supports Linux host and guest operating systems with kernel version 4.14 and 5.10. More information supported kernel versions can be found here. Firecracker also runs OSv guests

File: README.md [hunk 6]
```diff
@@ -141,6 +141,11 @@ Firecracker currently only supports little-endian platforms. Firecracker will
 not compile for big-endian architectures, and will not work correctly with
 big-endian configured guests.
 
+## Supported kernels
```
Comment (raduiliescu) [inline README.md:144]: I would name the version explicitly and then link for more details our support policy.

File: docs/getting-started.md [hunk 7]
```diff
@@ -23,11 +23,10 @@ sudo setfacl -m u:${USER}:rw /dev/kvm
 
 The generic requirements are explained below:
 
-- **Linux 4.14+**
+- **Linux**
 
   Firecracker currently supports physical Linux **x86_64** and **aarch64**
```
Comment (raduiliescu) [inline docs/getting-started.md:28]: I would put again the kernel versions explicitly, and *maybe* a reference to the support policy.

**Review by raduiliescu — COMMENTED** [review]
The policy looks good. I think it is better to be explicit in the FAQ, README and Getting started on the supported kernels. For example in the FAQ we should answer the question spot on, rather than providing a link. In design I am fine with leaving just a reference.

**Comment by dianpopa** [comment]
> The policy looks good. I think it is better to be explicit in the FAQ, README and Getting started on the supported kernels. For example in the FAQ we should answer the question spot on, rather than providing a link. In design I am fine with leaving just a reference.

I understand your suggestion but we need to be aware that every time we modify our support policy (adding or removing support support for some kernel) we would need to modify this 3 files. This is prone to errors. @firecracker-microvm/compute-capsule what is your take on this?

**Comment by georgepisaltu** [comment]
> > The policy looks good. I think it is better to be explicit in the FAQ, README and Getting started on the supported kernels. For example in the FAQ we should answer the question spot on, rather than providing a link. In design I am fine with leaving just a reference.
> 
> I understand your suggestion but we need to be aware that every time we modify our support policy (adding or removing support support for some kernel) we would need to modify this 3 files. This is prone to errors. @firecracker-microvm/compute-capsule what is your take on this?

Our documentation is constantly growing. I think referencing other parts of the documentation scales better than duplicating information throughout multiple files. I quite like the link approach in the FAQ for example, especially because kernel support is a complex topic and if a user is interested in it, they should in fact read the full documentation. The link not only makes the codebase more maintainable, but it also provides users who are interested in this an easier way of finding relevant documentation.

**Comment by alindima** [comment]
> > > The policy looks good. I think it is better to be explicit in the FAQ, README and Getting started on the supported kernels. For example in the FAQ we should answer the question spot on, rather than providing a link. In design I am fine with leaving just a reference.
> > 
> > 
> > I understand your suggestion but we need to be aware that every time we modify our support policy (adding or removing support support for some kernel) we would need to modify this 3 files. This is prone to errors. @firecracker-microvm/compute-capsule what is your take on this?
> 
> Our documentation is constantly growing. I think referencing other parts of the documentation scales better than duplicating information throughout multiple files. I quite like the link approach in the FAQ for example, especially because kernel support is a complex topic and if a user is interested in it, they should in fact read the full documentation. The link not only makes the codebase more maintainable, but it also provides users who are interested in this an easier way of finding relevant documentation.

I agree with Diana and George. I think this approach scales better and we'll most surely forget about this in the long run.
There are examples of such outdated documentation even on our Readme: https://github.com/firecracker-microvm/firecracker#features--capabilities

**Comment by raduiliescu** [comment]
Ok, let's push it like that.

**Review by raduiliescu — APPROVED** [review]

**Pull request merged**